File: deck/base/code/native/rust/runtime/atomic.rs
// Atomic integer runtime over std AtomicI64, shared through an Arc so the Clone-derived seed struct can hold it. Reached
// only through the public atomic API.
mod atomic {
    use std::sync::atomic::{AtomicI64, Ordering};
    use std::sync::Arc;
    pub fn make(initial: i64) -> Arc<AtomicI64> {
        Arc::new(AtomicI64::new(initial))
    }
    pub fn load(cell: Arc<AtomicI64>) -> i64 {
        cell.load(Ordering::SeqCst)
    }
    pub fn store(cell: Arc<AtomicI64>, value: i64) {
        cell.store(value, Ordering::SeqCst);
    }
    pub fn increase(cell: Arc<AtomicI64>, delta: i64) -> i64 {
        cell.fetch_add(delta, Ordering::SeqCst) + delta
    }
}

File: deck/base/code/native/node/runtime/atomic.ts
// Atomic integer runtime over Atomics on a SharedArrayBuffer-backed Int32Array, so the cell is safe to share with a
// worker thread. The opaque handle a seed atomic holds is the typed array. Reached only through the public atomic API.
const atomic = {
  make: (initial: number): Int32Array => {
    const cell = new Int32Array(new SharedArrayBuffer(4))
    Atomics.store(cell, 0, initial)
    return cell
  },
  load: (cell: Int32Array): number => Atomics.load(cell, 0),
  store: (cell: Int32Array, value: number): void => {
    Atomics.store(cell, 0, value)
  },
  increase: (cell: Int32Array, delta: number): number =>
    Atomics.add(cell, 0, delta) + delta,
}
